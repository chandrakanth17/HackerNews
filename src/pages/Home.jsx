import React from "react";
import { Helmet } from "react-helmet";

const head = () => {
  return (
    <Helmet>
      <title>My Page title</title>
    </Helmet>
  );
};

const Home = () => {
  return (
    <div>
      {head()}
      <h1>Home page</h1>
      <p>Some content</p>
      <button onClick={() => console.log("hello im running")}>
        Click the button
      </button>
    </div>
  );
};

export default Home;
