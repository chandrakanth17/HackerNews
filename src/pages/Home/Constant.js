export const RESPONSIVE = {
  mobileSm: `320px`,
  mobileMd: `375px`,
  mobileLg: `425px`,
  tablet: `768px`,
  desktopSm: `1024px`,
  desktopLg: `1440px`
};

export const INITIAL_LENGTH = 0;
