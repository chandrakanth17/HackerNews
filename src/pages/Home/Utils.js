import React from 'react';
import { Helmet } from 'react-helmet';
import moment from 'moment';

export const head = () => {
  return (
    <Helmet>
      <meta charSet="UTF-8" />
      <meta name="description" content="Y-combinator news clone" />
      <meta name="keywords" content="news, Y-combinator, hackernews" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>News feed Y-combinator</title>
    </Helmet>
  );
};

export const getDomain = url => {
  if (url) {
    const baseUrl = url.split('//')[1];
    return `(${baseUrl.split('/')[0]})`;
  }
  return 'blank';
};

export const timeFromNow = time => {
  const toTime = moment();
  return moment(toTime).from(time);
};

export const getHidden = () => {
  return JSON.parse(localStorage.getItem('hiddenArticles')) || [];
};

export const setHidden = hiddenArticles => {
  localStorage.setItem('hiddenArticles', JSON.stringify(hiddenArticles));
};

export const filterFeeds = (currentFeeds, hiddenArticles) => {
  return currentFeeds.filter(feed => {
    return !hiddenArticles.includes(feed.objectID);
  });
};
