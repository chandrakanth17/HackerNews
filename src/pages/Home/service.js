import axios from 'axios';
import { INITIAL_LENGTH } from './Constant';

export const getNewsFeed = (pageNumber = INITIAL_LENGTH) => {
  return axios
    .request({
      url: `http://hn.algolia.com/api/v1/search?tags=front_page&page=${pageNumber}`,
      method: 'get'
    })
    .then(response => {
      return response;
    })
    .catch(err => {
      console.log(err);
    });
};
