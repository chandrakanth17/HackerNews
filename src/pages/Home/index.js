import React, { useEffect, useState } from 'react';
import { getNewsFeed } from './service';
import { INITIAL_LENGTH } from './Constant';
import {
  NavBar,
  AppContainer,
  Container,
  Table,
  TableBody,
  MoreButton
} from './styles/FeedStyle';
import TableRowItem from './Components/TableRow';
import { head, getHidden, setHidden, filterFeeds } from './Utils';

const Home = () => {
  const [currentPage, setCurrentPage] = useState(0);
  const [currentFeeds, setCurrentFeeds] = useState([]);

  useEffect(async () => {
    const response = await getNewsFeed();
    if (response && response.data) {
      const filteredData = filterFeeds(response.data.hits, getHidden());
      await setCurrentPage(response.data.page);
      await setCurrentFeeds(filteredData);
    }
  }, []);

  const fetchNextPage = async () => {
    const response = await getNewsFeed(currentPage + 1);
    if (response && response.data) {
      await setCurrentFeeds([...currentFeeds, response.data.hits].flat());
      await setCurrentPage(response.data.page);
    }
  };
  const hideArticle = id => {
    const hiddenArticles = getHidden() || [];
    hiddenArticles.push(id);
    setHidden(hiddenArticles);
    const filteredFeeds = filterFeeds(currentFeeds, hiddenArticles);
    setCurrentFeeds(filteredFeeds);
  };

  const upVote = index => {
    const updatedFeeds = [...currentFeeds];
    updatedFeeds[index].points += 1;
    setCurrentFeeds(updatedFeeds);
  };

  return (
    <AppContainer>
      {head()}
      <NavBar>Y-Combinator</NavBar>
      {currentFeeds.length !== INITIAL_LENGTH ? (
        <Container>
          <Table cellspacing="0" cellpadding="0">
            <TableBody>
              {currentFeeds.map((feed, index) => {
                return (
                  <TableRowItem
                    key={index}
                    feed={feed}
                    upVote={() => upVote(index)}
                    hideArticle={() => hideArticle(feed.objectID)}
                  />
                );
              })}
            </TableBody>
          </Table>
          <MoreButton onClick={fetchNextPage}>More</MoreButton>
        </Container>
      ) : (
        <p>Loading...</p>
      )}
    </AppContainer>
  );
};

export default Home;
