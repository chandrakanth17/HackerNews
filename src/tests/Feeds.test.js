import React from 'react';
import 'babel-polyfill';
import { mount, configure, shallow } from 'enzyme';
import axios from 'axios';
import Home from '../pages/Home';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

jest.mock('axios');
const fnClick = jest.fn();

describe('It should display the list of feeds on feeds updates', () => {
  it('Loads the data on component mount and data gets fetches', () => {
    const App = renderer.create(<Home />).toJSON;
    expect(App).toMatchSnapshot();
  });

  it('It should validate feeds data when app is mounted', () => {
    const App = mount(<Home />);
    const Table = App.find('table');
    expect(Table).toBeTruthy();
  });
});
