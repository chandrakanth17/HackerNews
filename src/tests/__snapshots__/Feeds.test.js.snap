// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should display the list of feeds on feeds updates Loads the data on component mount and data gets fetches 1`] = `[Function]`;
